### PR TITLE
licensing: change access token format to 'slk_$hex($sha256($sha256(licenseKey)))'

### PR DESCRIPTION
**Pull request by bobheadxi**
Changes the license-based access token to take the sha256 of the license key twice:

```diff
- slk_$hex($sha256(licenseKey))
+ slk_$hex($sha256($sha256(licenseKey)))
```

This aligns with other hashing of sensitive things (license check, dotcom user tokens) and some Stackoverflow answers indicate this prevents [certain vulnerabilities](https://crypto.stackexchange.com/questions/50017/why-hashing-twice).

Until we migrate our existing instances over, our check retains back-compat with previous tokens - see the back-compat test.

Latency on token checks is not currently an issue, since we run full syncs periodically. We're [exploring pre-computing it](https://sourcegraph.slack.com/archives/C05B23Y3VRP/p1686595502071599).

## Test plan

Tests

### CODE REVIEW

**Review by eseliger — APPROVED** [review]

File: enterprise/cmd/frontend/internal/dotcom/productsubscription/tokens_db.go [hunk 1]
```diff
@@ -54,7 +54,10 @@ SELECT product_subscription_id
 FROM product_licenses
 WHERE
 	access_token_enabled=true
-	AND digest(license_key, 'sha256')=%s`,
+	AND ( digest(digest(license_key, 'sha256'), 'sha256')=%s OR digest(license_key, 'sha256')=%s )`,
+		decoded,
+		// retain backcompat for now, see 'legacy token format' test case
+		// needs migration: sourcegraph.com, k8s.sgdev.org, cody-dev, S2
```
Comment (eseliger) [inline enterprise/cmd/frontend/internal/dotcom/productsubscription/tokens_db.go:60]: dev-private now, too, I think

Comment (bobheadxi) [reply enterprise/cmd/frontend/internal/dotcom/productsubscription/tokens_db.go]: Added to the list!

**Review by mohammadualam — APPROVED** [review]

**Comment by evict** [comment]
@bobheadxi hashing the access token twice is unnecessary in this case. It won't add a lot of additional benefits because our token is randomly generated and 40 characters long. SHA2 is also a fine hashing algorithm. If you want additional security you could opt to choose a different algorithm such as Argon2.

**Review by unknwon — APPROVED** [review]

**Comment by bobheadxi** [comment]
@evict thanks, that's super useful context. I proposed this change in the context of https://github.com/sourcegraph/sourcegraph/pull/53331, where it seems [#job-fair-licensing-gating](https://sourcegraph.slack.com/archives/C05B23Y3VRP/p1686595502071599) added a new precomputed column that hashes the license key twice - I'm hoping to avoid having two variants of this logic and "access token", and thought we might unify on hashing it twice.

If hashing it once is sufficient and/or recommended, we can unify the two on hashing it once. I'll share in Slack and tag you in the conversation

**Pull request closed without merging**